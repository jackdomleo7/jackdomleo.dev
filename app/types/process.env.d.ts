declare namespace NodeJS {
  export interface ProcessEnv {
    NUXT_BASE_URL: string;
  }
}