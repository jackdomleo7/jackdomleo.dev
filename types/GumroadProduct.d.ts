export interface GumroadProduct {
  success: boolean;
  product: {
    name: string;
    preview_url: string;
    description: string;
    customizable_price: boolean;
    require_shipping: boolean;
    custom_receipt: string;
    custom_permalink: string;
    subscription_duration: null;
    id: string;
    url: string | null;
    price: number;
    currency: string;
    short_url: string;
    thumbnail_url: string | null;
    tags: string[];
    formatted_price: string;
    published: boolean;
    show_on_profile: boolean;
    file_info: Record<string, string>;
    max_purchase_count: number | null;
    deleted: boolean;
    custom_summary: string;
    is_tiered_membership: boolean;
    recurrences: null;
    variants: unknown[];
    release_at: string;
    sales_count: number;
    sales_usd_cents: number;
  }
}