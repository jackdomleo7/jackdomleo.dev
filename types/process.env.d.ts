declare namespace NodeJS {
  export interface ProcessEnv {
    NUXT_BASE_URL: string;
    NUXT_CTF_SPACE_ID: string;
    NUXT_CTF_CDA_ACCESS_TOKEN: string;
    NUXT_CTF_CDA_ACCESS_TOKEN_PREVIEW: string;
  }
}