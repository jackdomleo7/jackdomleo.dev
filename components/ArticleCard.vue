<template>
  <nuxt-link :to="`/blog${article.path}`" class="article-card">
    <nuxt-img :src="`/blog${article.path}/cover_image.png`" alt="" width="500" height="210" loading="lazy" class="article-card__img" />
    <div class="article-card__info">
      <ul v-if="article.tags" class="article-card__tags">
        <li v-for="tag in article.tags" :key="tag" class="tag--inverted">
          {{ tag }}
        </li>
      </ul>
      <component :is="headingLevel" class="article-card__heading">
        {{ article.title }}
      </component>
      <p class="article-card__description">
        {{ article.description }}
      </p>
      <div class="article-card__meta">
        <time :datetime="format(new Date(article.date), 'y-MM-dd')">{{ format(new Date(article.date), 'MMMM d, y') }}</time>
        <p>{{ article.readingTime }} min read</p>
      </div>
    </div>
  </nuxt-link>
</template>

<script lang="ts">
import Vue, { PropType } from 'vue'
import { format } from 'date-fns'
import { IArticle } from '@/types'

export default Vue.extend({
  name: 'ArticleCard',
  props: {
    article: {
      type: Object as PropType<IArticle>,
      required: true
    },
    headingLevel: {
      type: String as PropType<'h2'|'h3'|'h4'|'h5'|'h6'>,
      required: true,
      validator: (value: string): boolean => {
        return ['h2', 'h3', 'h4', 'h5', 'h6'].includes(value)
      }
    }
  },
  data () {
    return {
      format
    }
  }
})
</script>

<style lang="scss" scoped>
.article-card {
  display: flex;
  flex-direction: column;
  background-color: var(--colour-foreground-1);
  color: var(--colour-text-secondary);
  box-shadow: var(--shadow-standard);
  border-radius: 1.25rem;
  text-decoration: none;
  height: 100%;
  max-width: 31.25rem;
  overflow: hidden;

  &__img {
    width: 100%;
    background-color: #ccc;
    color: #000;
  }

  &__info {
    padding: 0.5rem 0.5rem 0.75rem 0.5rem;
    flex: 1;
    display: flex;
    flex-direction: column;

    @media (min-width: $responsive-small-tablet) {
      padding: 1rem 0.5rem 1rem 1rem;
    }
  }

  &__tags {
    padding-left: 0;
    margin-block: 0;
    list-style-type: none;
    display: flex;
    gap: 1rem;
  }

  &__heading {
    color: var(--colour-text-primary);
    font-size: var(--text-heading);
    font-weight: 400;
    margin-top: 0.5rem;
    margin-bottom: 0;

    @media (min-width: $responsive-small-tablet) {
      margin-top: 0.75rem;
    }
  }

  &__description {
    font-size: var(--text-body);
    margin-top: 1rem;
    margin-bottom: 0;
    flex: 1;

    @media (min-width: $responsive-small-tablet) {
      margin-top: 1.25rem;
    }
  }

  &__meta {
    display: flex;
    align-items: center;
    gap: 1rem;
    margin-top: 1rem;
    margin-bottom: 0;

    time {
      position: relative;

      &::after {
        content: 'Â·';
        position: absolute;
        right: -0.5rem;
        transform: translateX(50%);
      }
    }
  }
}
</style>
