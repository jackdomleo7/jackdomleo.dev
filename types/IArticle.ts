export interface INuxtContentGeneric {
  toc: {
    id: string;
    depth: number;
    text: string;
  }[];
  dir: string;
  path: string;
  extension: '.md';
  createdAt: string; // Not sure why Nuxt Content doesn't set this to a date
  updatedAt: string; // Not sure why Nuxt Content doesn't set this to a date
  body: any;
  excerpt: any;
  text: string;
}

export type TNuxtContentGenericRuntime = Omit<INuxtContentGeneric, 'excerpt'|'text'> // excerpt and text only exist during build time

export interface IArticle extends TNuxtContentGenericRuntime {
  slug: string;
  description: string;
  title: string;
  date: string; // In format of yyyy-mm-dd
  tags: string[];
  img: {
    url: string;
    alt: string;
  };
  readingTime: number; // Automatically calculated during build time
  embeds?: 'twitter'[]
}
