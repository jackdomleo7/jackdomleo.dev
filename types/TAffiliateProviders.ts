export type TAffiliateProviders = 'Amazon' | 'Gumroad'