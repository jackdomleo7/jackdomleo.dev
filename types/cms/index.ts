import { IAmazonLegal } from "./IAmazonLegal"
import { IPage } from "./IPage"
import { IPageHome } from "./IPageHome"
import { IPageAbout } from "./IPageAbout"
import { IPageProjects } from "./IPageProjects"
import { IPageBlog } from "./IPageBlog"
import { IPageProduct } from "./IPageProduct"

export {
  IAmazonLegal,
  IPage,
  IPageHome,
  IPageAbout,
  IPageProjects,
  IPageBlog,
  IPageProduct
}
