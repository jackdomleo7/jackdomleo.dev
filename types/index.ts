import { IArticle, INuxtContentGeneric, TNuxtContentGenericRuntime } from './IArticle'
import { TAffiliateProviders } from './TAffiliateProviders'

export {
  IArticle,
  INuxtContentGeneric,
  TNuxtContentGenericRuntime,
  TAffiliateProviders
}