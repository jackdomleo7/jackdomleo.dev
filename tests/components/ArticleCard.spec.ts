import { mount, shallowMount } from '@vue/test-utils'
import { configureAxe, toHaveNoViolations } from 'jest-axe'
import ArticleCard from '@/components/ArticleCard.vue'
import { IArticle } from '@/types'

const article: IArticle = {
  title: 'Some article about something cool',
  tags: ['frontend', 'a11y'],
  description: 'This is some really interesting article about something seriously friggin cool!',
  readingTime: 6,
  date: '2020-01-01',
  createdAt: String(new Date('2020-01-01')),
  updatedAt: String(new Date('2020-01-01')),
  body: {}, // Can be empty since we don't need the body in this component,
  dir: '/blog/',
  extension: '.md',
  slug: 'some-cool-article',
  path: '/blog/2020/some-cool-article',
  toc: [] // Can be empty since we don't use it in this component
}

expect.extend(toHaveNoViolations)

const axe = configureAxe({
  rules: {
    region: { enabled: false } // Don't need to check because components are loaded in isolation
  }
})

describe('ArticleCard', () => {
  /**
   * Vue instance
   */

  it('is a Vue instance', () => {
    const wrapper = mount(ArticleCard, {
      propsData: {
        article,
        headingLevel: 'h3'
      },
      stubs: ['nuxt-link', 'nuxt-img']
    })

    expect(wrapper.vm).toBeTruthy()
  })

  /**
   * Snapshots
   */

  it('matches snapshot when mounted', () => {
    const wrapper = mount(ArticleCard, {
      propsData: {
        article,
        headingLevel: 'h3'
      },
      stubs: ['nuxt-link', 'nuxt-img']
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('matches snapshot when shallow mounted', () => {
    const wrapper = shallowMount(ArticleCard, {
      propsData: {
        article,
        headingLevel: 'h3'
      },
      stubs: ['nuxt-link', 'nuxt-img']
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  /**
   * A11y
   */

  it('to have no obvious accessibility violations', async () => {
    const wrapper = mount(ArticleCard, {
      propsData: {
        article,
        headingLevel: 'h3'
      },
      stubs: ['nuxt-link', 'nuxt-img']
    })

    await expect(axe(wrapper.element)).resolves.toHaveNoViolations()
  })

  /**
   * Props
   */

  it('renders a <h2> when "h2" is set as the heading level', () => {
    const wrapper = shallowMount(ArticleCard, {
      propsData: {
        article,
        headingLevel: 'h2'
      },
      stubs: ['nuxt-link', 'nuxt-img']
    })

    expect(wrapper.find('h2.article-card__heading').exists()).toBe(true)
  })
})
