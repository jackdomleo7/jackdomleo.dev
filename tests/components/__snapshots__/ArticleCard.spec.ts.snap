// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleCard matches snapshot when mounted 1`] = `
<nuxt-link-stub
  class="article-card"
  to="/blog/blog/2020/some-cool-article"
>
  <nuxt-img-stub
    alt="A really cool thing"
    class="article-card__img"
    height="210"
    loading="lazy"
    src="/something-cool.png"
    width="500"
  />
   
  <div
    class="article-card__info"
  >
    <ul
      class="article-card__tags"
    >
      <li
        class="tag--inverted"
      >
        
        frontend
      
      </li>
      <li
        class="tag--inverted"
      >
        
        a11y
      
      </li>
    </ul>
     
    <h3
      class="article-card__heading"
    >
      
      Some article about something cool
    
    </h3>
     
    <p
      class="article-card__description"
    >
      
      This is some really interesting article about something seriously friggin cool!
    
    </p>
     
    <div
      class="article-card__meta"
    >
      <time
        datetime="2020-01-01"
      >
        January 1, 2020
      </time>
       
      <p>
        6 min read
      </p>
    </div>
  </div>
</nuxt-link-stub>
`;

exports[`ArticleCard matches snapshot when shallow mounted 1`] = `
<nuxt-link-stub
  class="article-card"
  to="/blog/blog/2020/some-cool-article"
>
  <nuxt-img-stub
    alt="A really cool thing"
    class="article-card__img"
    height="210"
    loading="lazy"
    src="/something-cool.png"
    width="500"
  />
   
  <div
    class="article-card__info"
  >
    <ul
      class="article-card__tags"
    >
      <li
        class="tag--inverted"
      >
        
        frontend
      
      </li>
      <li
        class="tag--inverted"
      >
        
        a11y
      
      </li>
    </ul>
     
    <h3
      class="article-card__heading"
    >
      
      Some article about something cool
    
    </h3>
     
    <p
      class="article-card__description"
    >
      
      This is some really interesting article about something seriously friggin cool!
    
    </p>
     
    <div
      class="article-card__meta"
    >
      <time
        datetime="2020-01-01"
      >
        January 1, 2020
      </time>
       
      <p>
        6 min read
      </p>
    </div>
  </div>
</nuxt-link-stub>
`;
